x/rollapp/keeper/block_height_to_finalization_queue.go

```diff
@@ -2,65 +2,114 @@ package keeper
 
 import (
 	"fmt"
+	"slices"
 
 	"github.com/cosmos/cosmos-sdk/store/prefix"
 	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/osmosis-labs/osmosis/v15/osmoutils"
+
 	common "github.com/dymensionxyz/dymension/v3/x/common/types"
 	"github.com/dymensionxyz/dymension/v3/x/rollapp/types"
-	"github.com/osmosis-labs/osmosis/v15/osmoutils"
 )
 
 // FinalizeRollappStates is called every block to finalize states when their dispute period over.
-func (k Keeper) FinalizeRollappStates(ctx sdk.Context) error {
+func (k Keeper) FinalizeRollappStates(ctx sdk.Context) {
 	if uint64(ctx.BlockHeight()) < k.DisputePeriodInBlocks(ctx) {
-		return nil
+		return
 	}
 	// check to see if there are pending  states to be finalized
 	finalizationHeight := uint64(ctx.BlockHeight() - int64(k.DisputePeriodInBlocks(ctx)))
 	pendingFinalizationQueue := k.GetAllFinalizationQueueUntilHeightInclusive(ctx, finalizationHeight)
 
-	return osmoutils.ApplyFuncIfNoError(ctx,
-		// we trap at this granularity because we want to avoid iterating inside the
-		// wrapped function (above), and we want the function to be atomic, so
-		// we cannot trap inside finalizePending
-		func(ctx sdk.Context) error {
-			return k.finalizePending(ctx, pendingFinalizationQueue)
-		})
+	k.FinalizeAllPending(ctx, pendingFinalizationQueue)
 }
 
-func (k Keeper) finalizePending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) error {
-	// Iterate over all the pending finalization queue
+func (k Keeper) FinalizeAllPending(ctx sdk.Context, pendingFinalizationQueue []types.BlockHeightToFinalizationQueue) {
+	// cache the rollapps that failed to finalize at current EndBlocker execution
+	failedRollapps := make(map[string]uint64)
+	// iterate over all the pending finalization height queues
 	for _, blockHeightToFinalizationQueue := range pendingFinalizationQueue {
-		// finalize pending states
-		for _, stateInfoIndex := range blockHeightToFinalizationQueue.FinalizationQueue {
-			stateInfo := k.MustGetStateInfo(ctx, stateInfoIndex.RollappId, stateInfoIndex.Index)
-			if stateInfo.Status != common.Status_PENDING {
-				panic(fmt.Sprintf("invariant broken: stateInfo is not in pending state: rollapp: %s: status: %s", stateInfoIndex.RollappId, stateInfo.Status))
-			}
-			stateInfo.Finalize()
-			// update the status of the stateInfo
-			k.SetStateInfo(ctx, stateInfo)
-			// update the LatestStateInfoIndex of the rollapp
-			k.SetLatestFinalizedStateIndex(ctx, stateInfoIndex)
-			// call the after-update-state hook
-			keeperHooks := k.GetHooks()
-			err := keeperHooks.AfterStateFinalized(ctx, stateInfoIndex.RollappId, &stateInfo)
-			if err != nil {
-				return fmt.Errorf("after state finalized: %w", err)
-			}
-
-			// emit event
-			ctx.EventManager().EmitEvent(
-				sdk.NewEvent(types.EventTypeStatusChange,
-					stateInfo.GetEvents()...,
-				),
-			)
+		k.finalizeQueueForHeight(ctx, blockHeightToFinalizationQueue, failedRollapps)
```

Comment:
seems to me like the you can have a state index of e.g rollapp abc being finalized after  previous state index of rollapp abc failed in previous queue. e.g :

1. queue1 = Queue{height: 10, stateIndexes: [{index:1, rollapp: abc}]}
2. queue2 = Queue{height: 11, stateIndexes: [{index:2, rollapp: abc}]} 

In that case, if e.g queue1 fails, it will not finalize the state index, but it seems like queue2 can be independently finalized (As we're trying to finalize each rollapp state index queue without consideration of previous rollapp queue results) vs failing all subsequent state index for that rollapp in the following queues.

_not sure if there is some other place in the code which will fail index2, due to index1 not being finalized, but I don't see it explicitly in this code._

Also I'd suggest a simple validation of only finalizing a state index, if the previous is also finalized (which can avoid all this cases) as currently the finalize function is the following:

```
func (s *StateInfo) Finalize() {
	s.Status = common.Status_FINALIZED
}
```

Reply:
[here ](https://github.com/dymensionxyz/dymension/blob/b159bbd3c001b297f5b8f7d5ba6bc02f8cb308be/x/rollapp/keeper/block_height_to_finalization_queue.go#L39) you can see that if the rollapp had previously failed, its state finalization will be skipped

[here](https://github.com/dymensionxyz/dymension/blob/b159bbd3c001b297f5b8f7d5ba6bc02f8cb308be/x/rollapp/keeper/block_height_to_finalization_queue.go#L46) the failure is recorded and will be kept for the duration of the current `EndBlocker` execution. 

this will result in the failed state changes to not be deleted from the queue and iterated over next iteration of `EndBlocker`.

Reply:
right. missed that. my bad.